Blockzilla/UIComponents/AutocompleteTextField.swift

```diff
@@ -1,166 +1,384 @@
-        selectedTextRange = textRange(from: beginningOfDocument, to: beginningOfDocument)
+    func setTextWithoutSearching(_ text: String) {
+        super.text = text
+        hideCursor = autocompleteTextLabel != nil
+        removeCompletion()
     }
 
-    private func applyCompletion() {
-        guard completionRange != nil else { return }
-
-        completionRange = nil
+    @objc func textDidChange(_ textField: UITextField) {
+        hideCursor = autocompleteTextLabel != nil
+        removeCompletion()
 
-        // Clear the current completion, then set the text without the attributed style.
-        // The attributed string must have at least one character to clear the current style.
-        let text = self.text ?? ""
-        attributedText = NSAttributedString(string: " ")
-        self.text = text
+        let isKeyboardReplacingText = lastReplacement != nil
+        if isKeyboardReplacingText, markedTextRange == nil {
+            notifyTextChanged?()
+        } else {
+            hideCursor = false
+        }
+    }
 
-        // Move the cursor to the end of the completion.
+    // Reset the cursor to the end of the text field.
+    // This forces `caretRect(for position: UITextPosition)` to be called which will decide if we should show the cursor
+    // This exists because ` caretRect(for position: UITextPosition)` is not called after we apply an autocompletion.
+    private func forceResetCursor() {
+        selectedTextRange = nil
         selectedTextRange = textRange(from: endOfDocument, to: endOfDocument)
     }
 
-    private func removeCompletion() {
-        guard let completionRange = completionRange else { return }
+    public override func deleteBackward() {
+        lastReplacement = ""
+        hideCursor = false
+        if isSelectionActive {
+            removeCompletion()
+            forceResetCursor()
+        } else {
+            super.deleteBackward()
+        }
+    }
 
+    public override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
         applyCompletion()
+        super.touchesBegan(touches, with: event)
+    }
+}
 
-        // Fixes: https://github.com/mozilla-mobile/focus-ios/issues/630
-        // Prevents the hard crash when you select all and start a new query
-        guard let count = text?.count, count > 1 else { return }
-
-        if count >= completionRange.location + completionRange.length {
-            text = (text as NSString?)?.replacingCharacters(in: completionRange, with: "")
+//extension AutocompleteTextField: MenuHelperInterface {
+//    override func canPerformAction(_ action: Selector, withSender sender: Any?) -> Bool {
+//        if action == MenuHelper.SelectorPasteAndGo {
+//            return UIPasteboard.general.hasStrings
+//        }
+//
+//        return super.canPerformAction(action, withSender: sender)
+//    }
+//
+//    @objc func menuHelperPasteAndGo() {
+//        autocompleteDelegate?.autocompletePasteAndGo(self)
+//    }
+//}
```

Comment:
same here: do we need this commented out code or can we remove it 🤔

Reply:
Yes. Actually since I took this from Firefox, I should see if I can make this code work.

Reply:
Removing the code for now. It seems that Paste and Go is implemented elsewhere.